### PR TITLE
Add test endpoint and ability to add tokens from backend

### PR DESCRIPTION
**Pull request by dshanske**
This adds two features

* Ability to add a custom token from the admin. This allows for it to be given to systems that can't support a token exchange workflow and for testing purposes. 

* A test endpoint built in for troubleshooting purposes. This is part of the attempt to diagnose problems with sites that aren't authenticating. The test endpoint basically tests the 'determine_current_user' flow by either returning an error or returning a successful token.

### CODE REVIEW

File: includes/class-indieauth-authorize.php [hunk 1]
```diff
@@ -15,12 +15,56 @@ public function __construct() {
...
 		add_action( 'send_headers', array( $this, 'http_header' ) );
 		add_action( 'wp_head', array( $this, 'html_header' ) );
 
 		add_filter( 'indieauth_scopes', array( $this, 'get_indieauth_scopes' ), 9 );
 		add_filter( 'indieauth_response', array( $this, 'get_indieauth_response' ), 9 );
+
+	}
+
+	// Test Route for Authorization
+	public static function register_routes() {
```
Comment (pfefferle) [inline includes/class-indieauth-authorize.php]: I wouldn't mix the test endpoint with the production system. I would encapsulate this in a separate class, so it is possible to enable/disable it via WP_DEBUG for example

File: includes/class-indieauth-token-ui.php [hunk 2]
```diff
@@ -40,6 +41,49 @@ public function admin_menu() {
 	public function options_callback() {
 	}
 
+	public function newtoken() {
```
Comment (pfefferle) [inline includes/class-indieauth-token-ui.php]: I would prefer `new_token` to be consistent with the rest of the code.

**Comment by dshanske** [comment]
@pfefferle I made the requested changes.

File: includes/class-indieauth-debug.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
+<?php
+
+class IndieAuth_Debug {
+
+
+	public function __construct() {
+		add_filter( 'http_request_args', array( $this, 'indieauth_allow_localhost' ), 10, 2 );
+		add_filter( 'rest_post_dispatch', array( $this, 'log_rest_api_errors' ), 10, 3 );
+				add_action( 'rest_api_init', array( $this, 'register_routes' ) );
```
Comment (pfefferle) [inline includes/class-indieauth-debug.php]: indents

File: includes/class-indieauth-debug.php [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+					get_current_user_id()
+				)
+			);
+		}
+		return $result;
+	}
+
+		// Test Route for Authorization
+	public static function register_routes() {
+			register_rest_route(
```
Comment (pfefferle) [inline includes/class-indieauth-debug.php]: indents

File: includes/class-indieauth-debug.php [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+						'callback'            => array( $this, 'test' ),
+						'permission_callback' => array( $this, 'permissions_test' ),
+					),
+				)
+			);
+
+	}
+
+	public static function permissions_test( $request ) {
+			$access_token = $request->get_param( 'access_token' );
```
Comment (pfefferle) [inline includes/class-indieauth-debug.php]: indents

File: includes/class-indieauth-debug.php [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+						'forbidden', __( 'No User is Currently Set', 'indieauth' ), array(
+							'status'  => 403,
+							'request' => $request,
+						)
+					);
+			} else {
+					return new WP_Error( 'forbidden', __( 'A Successful Auth Response was Given yet No User is Currently Set', 'indieauth' ), array( 'status' => 403 ) );
+			}
+		}
+			return true;
```
Comment (pfefferle) [inline includes/class-indieauth-debug.php]: indents

File: includes/class-indieauth-debug.php [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	public static function test( $request ) {
+		if ( 0 === get_current_user_id() ) {
+				return new WP_Error( 'forbidden', __( 'You have passed a permissions test but somehow the system is still not showing you as logged in', 'indieauth' ) );
+		}
+			return indieauth_get_response();
```
Comment (pfefferle) [inline includes/class-indieauth-debug.php]: indents

File: includes/class-indieauth-debug.php [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+			} else {
+					return new WP_Error( 'forbidden', __( 'A Successful Auth Response was Given yet No User is Currently Set', 'indieauth' ), array( 'status' => 403 ) );
+			}
+		}
+			return true;
+	}
+
+	public static function test( $request ) {
+		if ( 0 === get_current_user_id() ) {
+				return new WP_Error( 'forbidden', __( 'You have passed a permissions test but somehow the system is still not showing you as logged in', 'indieauth' ) );
```
Comment (pfefferle) [inline includes/class-indieauth-debug.php]: indents

**Review by pfefferle — APPROVED** [review]

**Comment by pfefferle** [comment]
cool, thanks.

**Pull request merged**